diff --git a/scripts/mod/modpost.c b/scripts/mod/modpost.c
index 3715db3..619b6f1 100644
--- a/scripts/mod/modpost.c
+++ b/scripts/mod/modpost.c
@@ -2089,11 +2089,11 @@ static void write_buf(struct buffer *b, const char *fname)
 	}
 }
 
-static void add_rhelversion(struct buffer *b, struct module *mod)
-{
-	buf_printf(b, "MODULE_INFO(rhelversion, \"%d.%d\");\n", RHEL_MAJOR,
-		   RHEL_MINOR);
-}
+//static void add_rhelversion(struct buffer *b, struct module *mod)
+//{
+//	buf_printf(b, "MODULE_INFO(rhelversion, \"%d.%d\");\n", RHEL_MAJOR,
+//		   RHEL_MINOR);
+//}
 
 static void write_if_changed(struct buffer *b, const char *fname)
 {
@@ -2155,7 +2155,7 @@ static void write_mod_c_file(struct module *mod)
 	add_depends(&buf, mod);
 	add_moddevtable(&buf, mod);
 	add_srcversion(&buf, mod);
-	add_rhelversion(&buf, mod);
+	//add_rhelversion(&buf, mod);
 
 	ret = snprintf(fname, sizeof(fname), "%s.mod.c", mod->name);
 	if (ret >= sizeof(fname)) {
