diff --git a/drivers/hid/hid-asus.c b/drivers/hid/hid-asus.c
index 3ae9a7d..8c45bc5 100644
--- a/drivers/hid/hid-asus.c
+++ b/drivers/hid/hid-asus.c
@@ -51,6 +51,7 @@ MODULE_DESCRIPTION("Asus HID Keyboard and TouchPad");
 #define FEATURE_KBD_REPORT_SIZE 16
 #define FEATURE_KBD_LED_REPORT_ID1 0x5d
 #define FEATURE_KBD_LED_REPORT_ID2 0x5e
+#define FEATURE_KBD_LED_REPORT_ID3 0x09
 
 #define SUPPORT_KBD_BACKLIGHT BIT(0)
 
@@ -454,6 +455,8 @@ static int rog_nkey_led_init(struct hid_device *hdev)
 				0x54, 0x65, 0x63, 0x68, 0x2e, 0x49, 0x6e, 0x63, 0x2e, 0x00 };
 	u8 buf_init3[] = { FEATURE_KBD_LED_REPORT_ID1,
 						0x05, 0x20, 0x31, 0x00, 0x08 };
+	//ROG ALLY LED init sequence
+	u8 buf_init4[] = { FEATURE_KBD_LED_REPORT_ID3, 0x5a, 0x03, 0x02, 0x00, 0x40, 0x00};
 	int ret;
 
 	hid_info(hdev, "Asus initialise N-KEY Device");
@@ -492,7 +495,12 @@ static int rog_nkey_led_init(struct hid_device *hdev)
 	ret = asus_kbd_set_report(hdev, buf_init3, sizeof(buf_init3));
 	if (ret < 0)
 		hid_warn(hdev, "Asus failed to send init command 2.1: %d\n", ret);
-
+	
+	//ROG ALLY LED init
+	ret = asus_kbd_set_report(hdev, buf_init4, sizeof(buf_init4));
+	if (ret < 0)
+		hid_warn(hdev, "Asus failed to send ROG ALLY LED init command: %d\n", ret);
+	
 	return ret;
 }
 
